crates/ide/src/annotations.rs

```diff
@@ -0,0 +1,139 @@
+use hir::Semantics;
+use ide_db::{
+    base_db::{FileId, FilePosition, FileRange, SourceDatabase},
+    RootDatabase, SymbolKind,
+};
+use syntax::TextRange;
+
+use crate::{
+    file_structure::file_structure,
+    fn_references::find_all_methods,
+    goto_implementation::goto_implementation,
+    references::find_all_refs,
+    runnables::{runnables, Runnable},
+    NavigationTarget, RunnableKind,
+};
+
+// Feature: Annotations
+//
+// Provides user with annotations above items for looking up references or impl blocks
+// and running/debugging binaries.
+pub struct Annotation {
+    pub range: TextRange,
+    pub kind: AnnotationKind,
+}
+
+pub enum AnnotationKind {
+    Runnable { debug: bool, runnable: Runnable },
+    HasImpls { position: FilePosition, data: Option<Vec<NavigationTarget>> },
+    HasReferences { position: FilePosition, data: Option<Vec<FileRange>> },
+}
+
+pub(crate) fn annotations(
+    db: &RootDatabase,
+    file_id: FileId,
+    binary_target: bool,
+    annotate_runnables: bool,
+    annotate_impls: bool,
+    annotate_references: bool,
+    annotate_method_references: bool,
+    run: bool,
+    debug: bool,
```

Comment:
these should be packed into a config struct: https://github.com/rust-analyzer/rust-analyzer/pull/7653/files